package com.patientmanagementsystem.patient.grpc;


import billing.BillingRequest;
import billing.BillingResponse;
import billing.BillingServiceGrpc;
import io.grpc.ManagedChannel;
import io.grpc.ManagedChannelBuilder;
import lombok.extern.slf4j.Slf4j;
import org.springframework.beans.factory.annotation.Value;
import org.springframework.stereotype.Service;

@Slf4j
@Service
public class BillingServiceGrpcClient {

    private final BillingServiceGrpc.BillingServiceBlockingStub blockingStub;

    public BillingServiceGrpcClient(
            @Value("${billing.service.address:localhost}") String serverAdress,
            @Value("${billing.service.grpc.port:9001}") int serverPort)
    {
        log.info("Connecting to billing service GRPC service at {}:{}", serverAdress,serverPort);
        ManagedChannel channel = ManagedChannelBuilder.forAddress(serverAdress,serverPort)
                .usePlaintext()
                .build();
        blockingStub = BillingServiceGrpc.newBlockingStub(channel);
    }

    public BillingResponse createBillingAccount(String patientId, String name, String email){
        BillingRequest request = BillingRequest.newBuilder().setPatientId(patientId)
                .setEmail(email)
                .setName(name)
                .build();
        BillingResponse response = blockingStub.createBillingAccount(request);
        log.info("Received response from billing GRPC service at {}", response);
        return  response;
    }

}
